### PR TITLE
[CORE] Introduce custom protobuf dependency with recursion limit enlarged

### PR DESCRIPTION
**Pull request by PHILO-HE**
## What changes were proposed in this pull request?

Protobuf has a hard-code limit (100) for recursion depth and runtime exception can occur when exceeding the limit. This PR proposes a custom protobuf dependency which has a much larger limit (100, 000, see source [here](https://github.com/oap-project/protobuf/tree/v3.23.4-0)). Thanks to @weiting-chen, the dependency jar has been published in maven repo (see [link](https://repo1.maven.org/maven2/io/glutenproject/protobuf-java/3.23.4-0/)).

Fixes: #1727, #2414. 

## How was this patch tested?

Existing tests.

### CODE REVIEW

File: gluten-core/pom.xml [hunk 1]
```diff
@@ -173,12 +173,14 @@
     <dependency>
       <groupId>com.google.protobuf</groupId>
       <artifactId>protobuf-java</artifactId>
-      <version>${protobuf.version}</version>
+      <version>${forked.protobuf.version}</version>
+      <scope>provided</scope>
```
Comment (PHILO-HE) [inline gluten-core/pom.xml:177]: We tried to replace this with custom protobuf, but got error (class symbol cannot be found) when building generated protobuf source. Maybe, there is some unknown requirement. So we just keep the original dependency in this module with "provided" scope used. The custom protobuf packed into gluten jar will be used at runtime.

**Comment by zhztheplayer** [comment]
Thanks for the fix @PHILO-HE. But had we tried using Java's reflection utilities to rewrite the recursion limit forcibly at runtime? I went through [code](https://github.com/oap-project/protobuf/commit/1597b3bc7c3c398e8c67aeddad5b8f76efa2f7f2#diff-0aa3b0d34acf2404473476e234080a2293dcdd26c8865d5f33454c1d1a246b46) then realized it might be a typical case that reflection is enough to cover.

**Comment by PHILO-HE** [comment]
> Thanks for the fix @PHILO-HE. But had we tried using Java's reflection utilities to rewrite the recursion limit forcibly at runtime? I went through [code](https://github.com/oap-project/protobuf/commit/1597b3bc7c3c398e8c67aeddad5b8f76efa2f7f2#diff-0aa3b0d34acf2404473476e234080a2293dcdd26c8865d5f33454c1d1a246b46) then realized it might be a typical case that reflection is enough to cover.

Hi @zhztheplayer, thanks for your suggestion! It should also work by using reflection to change the class static member in class's first loading. The proposed custom dependency is straightforward, but it requires some efforts to complete the maven publish process. Maybe, we can try to use reflection when protobuf is required to upgrade or needs some other similar changes in the future. Does it make sense to you?

**Review by zhztheplayer — APPROVED** [review]
Thanks for the explanation. I would be OK to both. Just thought that using some reflections could make things simpler.

**Pull request merged**